### PR TITLE
asyncio.locks: fix _ContextManagerMixin base class

### PR DESCRIPTION
**Pull request by hauntsaninja**

### CODE REVIEW

**Comment by hauntsaninja** [comment]
Hmm, this pattern was deprecated in 3.7 and removed in 3.9. See bpo-34793

**Comment by hauntsaninja** [comment]
Weirdly enough, just defining `__iter__` doesn't seem to be enough, you need to subclass for mypy to recognise it as iterable: https://github.com/python/mypy/blob/22e504b09d55adc024ecd19604367651b2865db1/mypy/checker.py#L2727

File: stdlib/3/asyncio/locks.pyi [hunk 1]
```diff
@@ -1,22 +1,34 @@
...
+        def __init__(self, lock: Union[Lock, Semaphore]) -> None: ...
+        def __aenter__(self) -> Awaitable[None]: ...
+        def __aexit__(self, exc_type: Optional[Type[BaseException]], exc: Optional[BaseException], tb: Optional[TracebackType]) -> Awaitable[None]: ...
+else:
+    class _ContextManager:
+        def __init__(self, lock: Union[Lock, Semaphore]) -> None: ...
+        def __enter__(self) -> object: ...
+        def __exit__(self, *args: Any) -> None: ...
+
+    class _ContextManagerMixin(Iterable[_ContextManager]):
```
Comment (JelleZijlstra) [inline stdlib/3/asyncio/locks.pyi:22]: I guess this was so you could do `with (yield from lock)` back in the dark ages of asyncio.

Comment (hauntsaninja) [reply stdlib/3/asyncio/locks.pyi:22]: Yeah, this is what that's about.

File: stdlib/3/asyncio/locks.pyi [hunk 2]
```diff
@@ -1,22 +1,34 @@
...
+        def __init__(self, lock: Union[Lock, Semaphore]) -> None: ...
+        # Apparently this exists to *prohibit* use as a context manager.
+        def __enter__(self) -> object: ...
+        def __exit__(self, *args: Any) -> None: ...
+        def __iter__(self) -> _ContextManager: ...
```
[the remark this answers is not in the document]
Comment (hauntsaninja) [reply stdlib/3/asyncio/locks.pyi]: In my non-CI attempt to repro this quickly, I started running into: `Cannot instantiate abstract class 'Condition' with abstract attribute '__iter__'`, which is why they're both there. Let's see what happens with the fixed return type. My non-CI repro doesn't bode well.

**Comment by JelleZijlstra** [comment]
The mypy selftest failure is actually concerning because it means the intended use case of these classes is no longer working correctly. Probably to fix this you need to duplicate `Future`'s iter annotation, something like `    def __iter__(self) -> Generator[Any, None, _T]: ...`. That will teach mypy that `(yield from lock)` returns a ContextManager.

**Comment by hauntsaninja** [comment]
Makes sense!

File: stdlib/3/asyncio/locks.pyi [hunk 3]
```diff
@@ -1,22 +1,34 @@
...
+        def __init__(self, lock: Union[Lock, Semaphore]) -> None: ...
+        def __enter__(self) -> object: ...
+        def __exit__(self, *args: Any) -> None: ...
+
+    class _ContextManagerMixin(Iterable[_ContextManager]):
+        def __init__(self, lock: Union[Lock, Semaphore]) -> None: ...
+        # Apparently this exists to *prohibit* use as a context manager.
+        def __enter__(self) -> object: ...
+        def __exit__(self, *args: Any) -> None: ...
+        def __iter__(self) -> Generator[bool, None, _ContextManager]: ...
```
Comment (JelleZijlstra) [inline stdlib/3/asyncio/locks.pyi:27]: The bool should be Any

**Pull request merged**